### ISSUE TITLE
Middleware state

### ISSUE THREAD

**Issue by garypretty** [Microsoft/botbuilder-dotnet#341]
Based on the conversations over in #339, I am opening this issue so that we can keep track and discuss the current issue of middleware not being able to maintain it's own state independently of the bot.

For the sake of clarity, the problem right now is that middleware is not able to store state in a de-coupled way from the bot itself (i.e. it would need to know about the structure of the POCO being used for bot state) and therefore this makes it impossible to build middleware that can be generic and used across multiple bots where the middleware requires some state management. Ideally the bot state needs to move to some similar form of collection with get / set (maybe still using POCOs - e.g. middleware would implement it's own POCO for state management) so that various pieces of the puzzle (middleware components and the bot) can store their own state object - probably with a named key for the cases where a piece of middleware is used multiple times on a bot and require a separate state object for each.

@billba over to you to provide some additional info based on your discussions I think? 👍

**Comment by billba** [comment]
This has always been possible, in that you can create your own storage key formula. For instance the following creates state scoped to a conversation, but specific to a namespace. (Obviously this is JS so please apply the necessary mental transformation for C#)

```ts
interface MyMiddlewareState {
  foo: string;
}

const MyMiddleware {
  static namespace = "github.com/billba/MyMWState";

  constructor(storage) {
    this.myState = new BotState<MyState>(storage, MyMiddleware.namespace, context =>
      `channel/${context.channelId}/conversation/${context.conversation.id}/${MyMiddleware.namespace}`
    );
  }

  async onProcessRequest(context, next) {
    const state = await this.myState.read(context);
    // maybe do stuff with state.foo
    await next();
    // maybe do stuff with state.foo
    await this.myState.write(context);
  }
}
```

Yesterday we decided to make this a little more turnkey. Specifically:
* All storage keys get a namespace suffix, e.g. `channel/{channelId}/conversation/{conversationId}/{namespace}`
* `new ConversationState()` gets a namespace argument (still to be decided: optional or required)

Please let me know if this solves the problem as you see it.

**Comment by garypretty** [comment]
@billba Ah, that wasn't my understanding of how state worked, at least in C# right now.  I thought that recent (within the last few weeks) changes were made to state that meant that essentially you could only have a single state object e.g. 'MyBotState' per bot.  If that is not the case then great, but we need to confirm.  Certainly when I last looked I could not see a way of achieving the above from an isolated perspective within middleware and that the only option was to add additional middleware specific properties to the 'MyBotState' object.  This is effectively my issue - if it is possible to have multiple state objects and add / create these within middleware then fantastic.

@cleemullins @drub0y Is the above example from @billba analogous with the C# SDK right now?

**Comment by billba** [comment]
It looks identical. Here's the relevant [code](https://github.com/Microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder.Core.Extensions/BotState.cs#L35) for .net

**Comment by billba** [comment]
Related JS conversation: https://github.com/Microsoft/botbuilder-js/issues/129

**Comment by garypretty** [comment]
@billba I totally see that I can add additional state objects, where I could make one specifically for my middleware. However, wouldn't that mean that I also needed to register my state middleware (given that BotState is itself a piece of middleware) prior to registering the middleware that used it? Otherwise at what point in the pipeline would the state be persisted?

**Comment by billba** [comment]
I made this same mistake myself before @Stevenic cleared it up for me.

*BotState can act as middleware, but it's purely optional.* You always have the ability to persist it yourself, and I did so above in the last line:
```ts
await this.myState.write(context);
```
I think this is the right pattern when middleware creates its own private state.

**Comment by garypretty** [comment]
@billba Wondering if this is where the divergence is between C# and JS.  Right now BotState is an abstract class in C# meaning like something like this is not possible from within middleware.  You can sub-class BotState, but the read / write methods are protected which mean that you cannot implement them on the sub-class.

```js
this.myState = new BotState<MyState>(storage, MyMiddleware.namespace, context =>
      `channel/${context.channelId}/conversation/${context.conversation.id}/${MyMiddleware.namespace}`
    );
```

It looks like the way ConversationState and UserState are implemented right now means they have to be registered and used as middleware in order to persist their state.


Again - I have been thinking about this for sometime now and I might just being going round in meaningless circles in my head, but I have tried to implement it in the way you suggested above and I just don't see how it can be done right now on the C# side.

**Comment by billba** [comment]
Those sound like bugs in the C# implementation.

**Comment by drub0y** [comment]
> What may be causing confusion is that there are some extension methods that hide this functionality.

Honestly I think it's the fact that you have to configure a middleware for each `TState` you intend to store  (and at each level you intend to store it) vs a single piece of middleware that enables state as a feature and then all the other state management APIs "light up" and let you store as many pieces of state of as many types as you want.

I think the changes we have discussed are going to greatly improve the experience here. The foundational work around services is in PR review now. Once that's in I'll be able to spike on the PoC for the changes to state management and we'll have some "real" code (vs whiteboard only) that demonstrates the full changes that are being proposed and start to have a real discussion around them... even if that leads to everyone scoffing at me in disgust. 😆

> You can sub-class BotState, but the read / write methods are protected which mean that you cannot implement them on the sub-class.

I don't think in the current API you would aim to subclass `BotState`, you would just new up a new subclass that targets which state you want to store and where... like new `ConversationState<MyState>`.

> BotState being abstract in C# sounds like a bug. Same with read and write being protected.

This stuff is (hopefully) going to change pretty substantially, so we might want to hold off improving what's there and make sure the new stuff looks right.

**Comment by garypretty** [comment]
@drub0y so just to confirm the case now, am I right in suggesting that in order to use the state functionality now in C# you do need to register some middleware and this cannot be done totally isolated within the middleware that requires state?

**Comment by billba** [comment]
I strongly feel we should fix what's there. They are both small fixes, and together they unblock @garypretty's scenario. Moreover we should implement the namespace feature as agreed upon (also a small change). Then if @drub0y's cunning plan comes to fruition, all the better.

**Comment by garypretty** [comment]
@drub0y Do you think I am right in saying that we can just make the BotState class non-abstract and make the read / write methods public vs protected to accomplish this?  If that's the case I would love to get this in as it will unblock a bunch of scenarios I want to progress in terms of some useful middleware. 

I could also add an optional string param which could be used as a namespace as @billba is suggesting.

**Comment by billba** [comment]
Yes. In my mind this is two PR's, one to bring C# BotState into alignment JS, another to add namespace support to [Converstion|User]State.

**Comment by garypretty** [comment]
@drub0y Maybe you can see if the above changes are really straightforward to make. i.e. allow you to do the following in C#.  I have just had a look and once I make the read / write public and remove the abstract from BotState I am mostly there, but then the Read method returns StoreItems and not StateT.

```js
interface MyMiddlewareState {
  foo: string;
}

const MyMiddleware {
  static namespace = "github.com/billba/MyMWState";

  constructor(storage) {
    this.myState = new BotState<MyState>(storage, MyMiddleware.namespace, context =>
      `channel/${context.channelId}/conversation/${context.conversation.id}/${MyMiddleware.namespace}`
    );
  }

  async onProcessRequest(context, next) {
    const state = await this.myState.read(context);
    // maybe do stuff with state.foo
    await next();
    // maybe do stuff with state.foo
    await this.myState.write(context);
  }
}
```

**Comment by garypretty** [comment]
Also - @billba - Thanks for working through this. Glad we finally figured out that right now what is possible in JS is not possible in C#. Getting this small change in will really help! :)

**Comment by billba** [comment]
That looks wrong to me. Here is the corresponding `read` in JS: https://github.com/Microsoft/botbuilder-js/blob/master/libraries/botbuilder-core-extensions/src/botState.ts#L53

**Comment by garypretty** [comment]
@billba Agreed. If nobody else gets a chance I'll take a look later (heading out for a few hours now) and see if I can at least bring the read / write into alignment.

**Comment by drub0y** [comment]
@garypretty Sorry, I have been working with a partner and haven't had a chance to keep up with everything going on. 😊

Yeah, so, I mean... your analysis of the current state is fair and everything you point out is absolutely true with the design choices that have been made up to this point. That said, I'm pretty sure thereis a way to work with it the way it is now. It's not great, but, yeah... you have to `.Use(...)` each instance of a new `ConversationState<T>` or `UserState<T>` for each `T` you want to store in each particular state "bucket". **Yes** it does mean the app has to register _N_ middleware for _all_ the `T`s have of state you might want and **yes** you have to live with the auto-saving, but it _SHOULD_ work because each one of those will be persisted into its own namespace based on the `T`.

As for changes, I personally have a hard time justifying spending any time on rev'ing the current design in any way because I know it's about to radically change to solve all these issues and enable all sorts of scenarios that we can't even do right now with the existing design.

**Comment by garypretty** [comment]
All totally makes sense @drub0y - Thanks for confirming my current understanding. If the changes to state are imminent then I understand not making further changes now. Thanks for working through though. All ultimately helps towards the final understanding /solution.

**Comment by billba** [comment]
@garypretty Just to be totally clear, there is no booked plan to make the changes @drub0y is discussing, but I greatly admire his confidence. I still think the right course is to bring C# into alignment with JS right now because it's hard to know how long it will take @drub0y to sell his plan.

**Comment by garypretty** [comment]
@billba That being the case, if it is potentially going to take some time, then I would agree that an interim solution of bringing things into line would be great.  I am happy to do the work if we can agree between us what needs doing. This will allow me to crack on a build some great generic middleware options for developers using v4 which can be packaged up and placed on NuGet.
@cleemullins @tomlm @Stevenic Can you guys weigh in on this please so that we can figure out what the best way forward is here?

**Comment by drub0y** [comment]
> but it SHOULD work because each one of those will be persisted into its own namespace based on the T.

Soooo, I just put this idea to the test and actually doesn't work because, while the state is stored under the services collection according to it's different namespace which includes the `T`:

https://github.com/Microsoft/botbuilder-dotnet/blob/18252f42a1bdcdb026acfe60d3e385e56fb7e811/libraries/Microsoft.Bot.Builder.Core.Extensions/BotState.cs#L94

the so called `keySelector` implementation for `UserState<T>` and `ConversationState<T>` actually always returns a singular string for all instances _regardless of `T`_: 

https://github.com/Microsoft/botbuilder-dotnet/blob/18252f42a1bdcdb026acfe60d3e385e56fb7e811/libraries/Microsoft.Bot.Builder.Core.Extensions/BotState.cs#L98

So in order to even use the idea I was proposing you'd need to _at least_ bug fix that.

**Comment by billba** [comment]
Right that's why I specifically suggested using `BotState`.

**Comment by garypretty** [comment]
@drub0y @billba Ok, this is only an initial stab at a couple of small alterations to allow BotState to be used as middleware via ConversationState / UserState or on it's own.  Take a look at https://github.com/garypretty/botbuilder-dotnet/blob/ae698e2087861d6e20edcbdb7cba41cfe727265c/libraries/Microsoft.Bot.Builder.Core.Extensions/BotState.cs

**Comment by billba** [comment]
Nice. That sample code looks right to me.

**Comment by drub0y** [comment]
It was low enough hanging fruit to fix without completely breaking the current model, so seems reasonable to me! I say PR it with concise explanation. 👍

**Comment by garypretty** [comment]
Closing this so we can keep conversation in the PR / DCR raised by @drub0y

**Comment by drub0y** [comment]
I want to reopen this issue since the State Management DCR PR (#376) has been put on hold.

**Comment by cleemullins** [comment]
This has been resolvbed with the merging of #345
